java/vector/src/main/java/org/apache/arrow/vector/dictionary/DictionaryHashTable.java

```diff
@@ -135,7 +139,8 @@ public int getIndex(int indexInArray, ValueVector toEncode) {
     for (DictionaryHashTable.Entry e = table[index]; e != null ; e = e.next) {
       if ((e.hash == hash)) {
         int dictIndex = e.index;
-        if (dictionary.equals(dictIndex, toEncode, indexInArray)) {
+        visitor.reset(indexInArray, dictIndex, 1);
+        if (toEncode.accept(visitor)) {
```

Comment:
Should we hide the fact that the impl uses a visitor from the APIs ? Maybe,

matcher = new RangeMatcher(srcVector, dstVector);
if (matcher.match(srcIndex, dstIndex, 1)) {
  return dictIndex;
}

Reply:
Sounds good, I add a ValueMatcher to do this. As the same reason mentioned above, this matcher only holds dictionary vector instead both.